Modules/_threadmodule.c

```diff
@@ -39,6 +40,127 @@ get_thread_state(PyObject *module)
     return (thread_module_state *)state;
 }
 
+// _ThreadHandle type
+
+typedef struct {
+    PyObject_HEAD
+    unsigned long long ident;
+    Py_uintptr_t handle;
+    char joinable;
+} ThreadHandleObject;
+
+static ThreadHandleObject*
+new_thread_handle(thread_module_state* state)
+{
+    ThreadHandleObject* self = PyObject_New(ThreadHandleObject, state->thread_handle_type);
+    if (self == NULL) {
+        return NULL;
+    }
+    self->ident = 0;
+    self->handle = 0;
+    self->joinable = 0;
+    return self;
+}
+
+static void
+ThreadHandle_dealloc(ThreadHandleObject *self)
+{
+    PyObject *tp = (PyObject *) Py_TYPE(self);
+    if (self->joinable) {
+        int ret = PyThread_detach_thread(self->handle);
+        if (ret) {
+            PyErr_SetString(ThreadError, "Failed detaching thread");
+            PyErr_WriteUnraisable(tp);
+        }
+    }
+    PyObject_Free(self);
+    Py_DECREF(tp);
```

Comment:
\<snark>
Ah heap types, how we "love" the ways you are different from static types and make it easy to leak references!
\</snark>

Reply:
Ah, I just grepped for `PyType_FromSpec` until I found a sufficiently similar example to copy and paste from :-)

Reply:
To be clear, I was not criticizing your code.  I was only lamenting how heap types have a few quirks that can easily trip people up, like forgetting to decref the type (which you didn't have to do for instances of static types). 😄